### PR TITLE
#202 Project Table Filter Component

### PR DESCRIPTION
**Pull request by Ari-S-123**
We implemented and tested all the features for the project table filter component we built for ticket #202. 
Screenshot (706)

### CODE REVIEW

File: src/components/projects/projects-table/projects-table.module.css [hunk 1]
```diff
@@ -3,6 +3,30 @@
  * See the LICENSE file in the repository root folder for details.
  */
 
-.table {
-  padding: 0em 2em 0em 2em;
+.container {
+  width: 100%;
+  padding-right: 15px;
```
Comment (jamescd18) [inline src/components/projects/projects-table/projects-table.module.css:8]: you might be able to shorten this with the single line padding that goes something like `padding: 5px 5px 5px 5px`

File: src/components/projects/projects-table/projects-table.test.tsx [hunk 2]
```diff
@@ -82,7 +89,7 @@ describe('projects table component', () => {
     ).toBeInTheDocument();
     expect(screen.getByText(wbsPipe(exampleAllProjects[4].wbsNum))).toBeInTheDocument();
 
-    expect(screen.getByText('All Projects')).toBeInTheDocument();
```
Comment (jamescd18) [inline src/components/projects/projects-table/projects-table.test.tsx:85]: Good catch!

File: src/components/projects/projects-table/projects-table.tsx [hunk 3]
```diff
@@ -10,9 +10,71 @@ import { DisplayProject } from './projects-table/projects-table';
...
+  const yearCheck = (project: Project) => {
+    return `${project.dateCreated.getUTCFullYear()}` === year;
+  };
+  const leadCheck = (project: Project) => {
+    return fullNamePipe(project.projectLead) === projectLead;
+  };
+  const managerCheck = (project: Project) => {
+    return fullNamePipe(project.projectManager) === projectManager;
+  };
+  if (carNumber != '') {
```
Comment (jamescd18) [inline src/components/projects/projects-table/projects-table.tsx:51]: use double equals whenever appropriate. here the linter complains

File: src/components/projects/projects-table/projects-table.tsx [hunk 4]
```diff
@@ -33,7 +95,90 @@ const ProjectsTable: React.FC = () => {
...
+  };
+
+  /**
+   * Returns an array of user names who are listed as a project's lead.
+   */
+  const getLeads = (): string[] => {
+    const projects = data!;
+    let leads: string[] = [];
+    for (let project of projects) {
+      let name = fullNamePipe(project.projectLead);
```
Comment (jamescd18) [inline src/components/projects/projects-table/projects-table.tsx:127]: very soon the `project.projectLead` field will be an array

File: src/components/projects/projects-table/projects-table.tsx [hunk 5]
```diff
@@ -33,7 +95,90 @@ const ProjectsTable: React.FC = () => {
...
+    projectLead,
+    projectManager
+  );
+  const projectLeads: string[] = getLeads();
+  const projectManagers: string[] = getManagers();
+
+  return (
+    <>
+      <div className={'mx-4 pt-1 pb-1'}>
+        <h3>Projects</h3>
```
Comment (jamescd18) [inline src/components/projects/projects-table/projects-table.tsx:164]: Pretty sure you should be using the page title component here

File: src/components/projects/projects-table/projects-table/projects-table.module.css [hunk 6]
```diff
@@ -4,5 +4,4 @@
  */
 
 .table {
```
Comment (jamescd18) [inline src/components/projects/projects-table/projects-table/projects-table.module.css:6]: I don't think you should leave empty CSS definitions. Having a CSS file which contains nothing other than the license comment at the top is totally fine

File: src/components/projects/projects-table/projects-table.tsx [hunk 7]
```diff
@@ -33,7 +95,90 @@ const ProjectsTable: React.FC = () => {
...
+    for (let project of projects) {
+      let name = fullNamePipe(project.projectManager);
+      if (!managers.includes(name)) {
+        managers.push(name);
+      }
+    }
+    return managers;
+  };
+
+  const filtered_data: Project[] = filterProjects(
```
Comment (jamescd18) [inline src/components/projects/projects-table/projects-table.tsx:150]: why do it here instead of inline on line 176?

**Review by jamescd18 — CHANGES_REQUESTED** [review]
This is a really great start! I've highlighted the things that stood out to me the most, but this will definitely need a couple more rounds of review and discussion before it's ready to merge. It's really awesome to see you two learning a lot and going beyond your comfort zone to build the component with state management in mind from the ground up. a few things:
- The year created is no longer relevant so that whole field and everything associated should be removed. sorry I didn't catch that and have it removed from the designs sooner
- are all the test IDs absolutely necessary for the testing? remember that the philosophy of testing react components with this testing library is "test what the user can see" and the user can't see test IDs. test IDs are effectively escape hatches to get away from that philosophy.
- remember to consider the naming of variables so that it's intelligible and makes sense
- remember to check the results of the linter (displays inline in the PR code changes tab)
- are there small pieces of abstraction that can be done? like you have a ton of dropdown items for "none", and the dropdown items for car number could perhaps be rewritten as `[1, 2].map(num => <Dropdown.Item ...... >)`?

As always, reply to inline comments, resolve inline comments, start a conversation on slack, and/or ping me when it's ready for another review! Great work so far!!

File: src/components/projects/projects-table/projects-table.tsx (hunk 4 above)
Comment (Ari-S-123) [reply src/components/projects/projects-table/projects-table.tsx:119]: I think we will have to wait until that other branch that deals with that is merged to make the changes in this line.

File: src/components/projects/projects-table/projects-table/projects-table.tsx [hunk 8]
```diff
@@ -3,15 +3,14 @@
  * See the LICENSE file in the repository root folder for details.
  */
 
-import { useHistory } from 'react-router-dom';
+import {useHistory} from 'react-router-dom';
```
Comment (jamescd18) [inline src/components/projects/projects-table/projects-table/projects-table.tsx:6]: pretty sure that if you have prettier enabled this change shouldn't happen

File: src/components/projects/projects-table/projects-table/projects-table.tsx [hunk 9]
```diff
@@ -3,15 +3,14 @@
...
 import BootstrapTable, {
   ColumnDescription,
   RowEventHandlerProps,
   SortOrder
 } from 'react-bootstrap-table-next';
-import PageTitle from '../../../shared/page-title/page-title';
 import styles from './projects-table.module.css';
-import { validateWBS } from 'utils';
+import {validateWBS} from 'utils';
```
Comment (jamescd18) [inline src/components/projects/projects-table/projects-table/projects-table.tsx:13]: same. shouldn't happen

File: src/custom.scss [hunk 10]
```diff
@@ -0,0 +1,7 @@
+// change the theme
+$theme-colors: (
+  'ner-red': #ef4345
+);
+
+// Import Bootstrap and its default variables
+@import '~bootstrap/scss/bootstrap.scss';
```
Comment (jamescd18) [inline src/custom.scss:7]: why? why on line 7?

File: src/custom.scss [hunk 11]
```diff
@@ -0,0 +1,7 @@
+// change the theme
```
Comment (jamescd18) [inline src/custom.scss:1]: needs license comment

File: src/index.tsx [hunk 12]
```diff
@@ -3,11 +3,11 @@
  * See the LICENSE file in the repository root folder for details.
  */
 
+import './custom.scss';
```
Comment (jamescd18) [inline src/index.tsx:6]: why make it the first one?

File: src/components/projects/projects-table/projects-table.tsx [hunk 13]
```diff
@@ -5,14 +5,67 @@
...
+    return carNumber === project.wbsNum.car.toString();
+  };
+  const statusCheck = (project: Project) => {
+    return project.status === status;
+  };
+  const leadCheck = (project: Project) => {
+    return listPipe(project.projectLead, fullNamePipe).includes(projectLead);
+  };
+  const managerCheck = (project: Project) => {
+    return fullNamePipe(project.projectManager) === projectManager;
```
Comment (jamescd18) [inline src/components/projects/projects-table/projects-table.tsx:44]: why string name equality when you really want ID equality?

File: src/components/projects/projects-table/projects-table.tsx [hunk 14]
```diff
@@ -5,14 +5,67 @@
...
+  projectManager: string
+): Project[] {
+  const carNumCheck = (project: Project) => {
+    return carNumber === project.wbsNum.car.toString();
+  };
+  const statusCheck = (project: Project) => {
+    return project.status === status;
+  };
+  const leadCheck = (project: Project) => {
+    return listPipe(project.projectLead, fullNamePipe).includes(projectLead);
```
Comment (jamescd18) [inline src/components/projects/projects-table/projects-table.tsx:41]: same question, why string equality? it should be ID-based

File: src/components/projects/projects-table/projects-table.tsx [hunk 15]
```diff
@@ -33,7 +86,80 @@ const ProjectsTable: React.FC = () => {
...
+          <div className={styles.filterTable}>
+            <ProjectsTableFilter
+              onClick={sendDataToParent}
+              leads={getLeads()}
+              managers={getManagers()}
+            />
+          </div>
+          <div className={styles.projectsTable}>
+            <PrjsTable
+              allProjects={transformToDisplayProjects(
```
Comment (jamescd18) [inline src/components/projects/projects-table/projects-table.tsx:154]: why not filter first and then transform? that would enable you to have more info for when filtering

File: src/components/projects/projects-table/projects-table/projects-table.tsx [hunk 16]
```diff
@@ -122,23 +121,22 @@ const ProjectsTable: React.FC<DisplayProjectProps> = ({ allProjects }: DisplayPr
...
-        keyField="wbsNum"
-        data={allProjects}
-        columns={columns}
-        defaultSorted={defaultSort}
-        rowEvents={rowEvents}
-        noDataIndication="No Projects to Display"
-        rowStyle={{ cursor: 'pointer' }}
-      />
-    </>
+      <>
```
Comment (jamescd18) [inline src/components/projects/projects-table/projects-table/projects-table.tsx:124]: might be able to remove these now that it's just one element inside

File: src/components/projects/projects-table/projects-table-filter/projects-table-filter.tsx [hunk 17]
```diff
@@ -0,0 +1,141 @@
+/*
+ * This file is part of NER's PM Dashboard and licensed under GNU AGPLv3.
+ * See the LICENSE file in the repository root folder for details.
+ */
+
+import { Button, Card, Dropdown, Form } from 'react-bootstrap';
+import styles from './projects-table-filter.module.css';
+import { useState } from 'react';
```
Comment (jamescd18) [inline src/components/projects/projects-table/projects-table-filter/projects-table-filter.tsx:8]: please put this import line first

**Review by jamescd18 — CHANGES_REQUESTED** [review]
This is pretty good! There's a design pattern question I included about filtering first instead of transforming to display projects first. I think filtering first is probably better. Also, I'd love it if you could explain more about why sass is required for just one button. A few more minor items here and there, and I'd love to see new screenshots added to a new comment in the PR.

**Comment by shanejpark** [comment]
> This is pretty good! There's a design pattern question I included about filtering first instead of transforming to display projects first. I think filtering first is probably better. Also, I'd love it if you could explain more about why sass is required for just one button. A few more minor items here and there, and I'd love to see new screenshots added to a new comment in the PR.

The only way to have custom colors for bootstrap objects is to either use !important or sass. The custom color in the sass file is useful if we're going to use more bootstrap objects later on or plan to modify them in any other way. There already are a lot of bootstrap elements in our project and it would be better if all of them just had a "variant: ner-red" line instead of having custom css or using !important.

File: src/custom.scss (hunk 10 above)
Comment (shanejpark) [reply src/custom.scss:7]: You have to modify the variables before importing, or else they'll be loaded without any modification

File: package.json [hunk 18]
```diff
@@ -115,7 +115,7 @@
     "@types/react-bootstrap-table-next": "^4.0.10",
     "@types/react-dom": "^16.9.10",
     "@types/react-router-dom": "^5.1.7",
-    "axios": "^0.21.1",
+    "axios": "^0.21.4",
```
Comment (jamescd18) [inline package.json:118]: Don't update versions unless justified / required as it could cause mysterious problems

File: package.json [hunk 19]
```diff
@@ -128,6 +128,7 @@
     "react-router-dom": "^5.2.0",
     "react-scripts": "4.0.1",
     "rollup": "^2.38.5",
+    "sass": "^1.43.4",
```
Comment (jamescd18) [inline package.json:131]: so is the verdict that yes you do want to use sass?

File: src/custom.scss [hunk 20]
```diff
@@ -0,0 +1,12 @@
+/*
+ * This file is part of NER's PM Dashboard and licensed under GNU AGPLv3.
+ * See the LICENSE file in the repository root folder for details.
+ */
+
+// change the theme
+$theme-colors: (
+  'ner-red': #ef4345
```
Comment (jamescd18) [inline src/custom.scss:8]: if you're gonna add one, please add the others from [the Brand Guideline doc](https://docs.google.com/document/d/1bV-TGorNeGbn7BhrwP9_nhSb_qmfs5wxJeJY5AfekDo/edit)

**Review by jamescd18 — CHANGES_REQUESTED** [review]
So the verdict is to go ahead and use sass?

File: src/components/projects/projects-table/projects-table.tsx [hunk 21]
```diff
@@ -33,7 +99,82 @@ const ProjectsTable: React.FC = () => {
...
+    return leads;
+  };
+
+  /**
+   * Returns an array of Users who are listed as a project's managers.
+   */
+  const getManagers = (): User[] => {
+    const projects = data!;
+    const managers: User[] = [];
+    const checklist: number[] = [];
```
Comment (jamescd18) [inline src/components/projects/projects-table/projects-table.tsx:145]: I think "seenList" would be more descriptive

**Review by jamescd18 — CHANGES_REQUESTED** [review]

File: src/components/projects/projects-table/projects-table.tsx [hunk 22]
```diff
@@ -5,14 +5,67 @@
...
+}
+
+/**
+ * Parent component for the projects page housing the filter table and projects table.
+ */
 const ProjectsTable: React.FC = () => {
+  const [status, setStatus] = useState('');
+  const [projectLead, setProjectLead] = useState('');
+  const [projectManager, setProjectManager] = useState('');
+  const [carNumber, setCarNumber] = useState('');
```
[the remark this answers is not in the document]
Comment (Ari-S-123) [reply src/components/projects/projects-table/projects-table.tsx]: Those changes seem to be old and don't reflect the current state of that file.

**Review by jamescd18 — APPROVED** [review]
This is sooooooooooo awesome to finally see completed! Amazing job guys! I love to see the perseverance and dedication to refining your development skills through trial and error here. Great work!

**Pull request merged**